compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -238,18 +238,38 @@ data ConversionEnv = ConversionEnv
   , convRange :: !(Maybe SourceLoc)
   }
 
-newtype ConvertM a = ConvertM (ReaderT ConversionEnv (Except FileDiagnostic) a)
-  deriving (Functor, Applicative, Monad, MonadError FileDiagnostic, MonadReader ConversionEnv)
+data ConversionState = ConversionState
+    { freshTmVarCounter :: Int
+    , freshTyVarCounter :: Int
+    }
+
+newtype ConvertM a = ConvertM (ReaderT ConversionEnv (StateT ConversionState (Except FileDiagnostic)) a)
```

Comment:
That’s quite a monad stack :slightly_smiling_face:

Reply:
Yeah, I know. But what can we do about it in a language like Haskell? We need all these effects. We could put the `ConversionEnv` in the `ConversionState`, but I'm not sure that makes the situation better.

Reply:
I don’t have an issue with it, I just got slightly amused :slightly_smiling_face: